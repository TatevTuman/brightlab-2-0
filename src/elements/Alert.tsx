import React, { memo } from 'react'
import { AlertComponentPropsWithStyle } from 'react-alert'
import cls from 'classnames'
import { ClassName } from '~types'

export interface AlertProps extends Omit<AlertComponentPropsWithStyle, 'id'> {
  className?: ClassName
}

// the style contains only the margin given as offset
// options contains all alert given options
// message is the alert message
// close is a function that closes the alert
const Alert: React.FC<AlertProps> = props => {
  const { className, style, options = {}, message, close } = props
  const { type } = options

  return (
    <div
      className={cls('w-fit min-h-40 flex font-medium items-center justify-center rounded-5 px-30', className)}
      data-type={type}
      style={style}
      onClick={close}
      data-testid={'alert'}
    >
      {message}
    </div>
  )
}

Alert.defaultProps = {
  style: {},
  message: 'Alert message',
  close: () => null,
  options: {
    onClose: () => null,
    onOpen: () => null,
    timeout: 2000
  }
}

export default memo(Alert)
