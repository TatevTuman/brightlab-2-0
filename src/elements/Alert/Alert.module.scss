.alert {
  width: fit-content;
  min-height: 44px;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 5px;
  padding: 0 30px;

  &[data-type="success"] {
    background: var(--primary);
    color: var(--white);
  }

  &[data-type="error"] {
    background: var(--danger);
    color: var(--white);
  }

  &-times {
    position: absolute;
    top: 10px;
    right: 5px;
  }
}
