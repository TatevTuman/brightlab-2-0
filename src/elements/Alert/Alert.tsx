import React from 'react'
import { AlertComponentPropsWithStyle } from 'react-alert'
import styles from './Alert.module.scss'

interface AlertProps extends AlertComponentPropsWithStyle {}

// the style contains only the margin given as offset
// options contains all alert given options
// message is the alert message
// close is a function that closes the alert
const Alert: React.FC<AlertProps> = props => {
  const { style, options, message, close } = props

  return (
    <div className={styles.alert} data-type={options.type} style={style}>
      {message}
    </div>
  )
}

export default Alert
