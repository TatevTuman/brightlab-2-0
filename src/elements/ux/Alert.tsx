import React, { memo } from 'react'
import { AlertComponentProps, AlertType } from 'react-alert'
import cls from 'classnames'
import { ClassName } from '~types'

export type AlertVariantType = AlertType
const getAlertVariantClass = (variant?: AlertVariantType, className?: ClassName) =>
  cls(
    'w-fit min-h-40 flex font-medium items-center justify-center rounded-5 px-30',
    {
      'bg-green-700': !variant || variant === 'success',
      'bg-orange-500': variant === 'info',
      'bg-red-700': variant === 'error'
    },
    className
  )

export interface AlertProps extends Omit<AlertComponentProps, 'id'> {
  className?: ClassName
}

// the style contains only the margin given as offset
// options contains all alert given options
// message is the alert message
// close is a function that closes the alert
const Alert: React.FC<AlertProps> = props => {
  const { className, options = {}, message, close } = props
  const { type: variant } = options

  return (
    <div className={getAlertVariantClass(variant, className)} onClick={close} data-testid={'alert'}>
      {message}
    </div>
  )
}

Alert.defaultProps = {
  message: 'Alert message',
  close: () => null,
  options: {
    onClose: () => null,
    onOpen: () => null,
    timeout: 2000
  }
}

export default memo(Alert)
