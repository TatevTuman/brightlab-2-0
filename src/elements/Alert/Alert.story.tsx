import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import { action } from '@storybook/addon-actions'
import { Provider as AlertProvider, transitions, useAlert, AlertProviderProps } from 'react-alert'
import Alert, { AlertProps } from './Alert'
import { Button } from '@elements'
import { storybook } from '@utils'

interface AlertStoryProps extends Omit<AlertProviderProps, 'template'> {
  type: AlertProps['options']['type']
  onClose: AlertProps['options']['onClose']
  onOpen: AlertProps['options']['onOpen']
  message: AlertProps['message']
  close: AlertProps['close']
  style: AlertProps['style']
}

const ALERT_PROVIDER_POSITIONS = [
  'top left',
  'top center',
  'top right',
  'middle left',
  'middle',
  'middle right',
  'bottom left',
  'bottom center',
  'bottom right'
] as const

export default {
  title: 'elements/Alert',
  argTypes: {
    message: { control: 'text', name: 'Message' },
    position: {
      control: {
        type: 'select',
        options: ALERT_PROVIDER_POSITIONS
      },
      name: 'Position'
    },
    type: { control: 'text', name: 'Message' },
    transition: { control: { type: 'select', options: ['fade', 'scale'] }, name: 'Transition' },
    timeout: { control: 'number', name: 'Timeout' },
    offset: { control: 'text', name: 'Offset' },
    style: { control: 'object', name: 'Style' },
    onOpen: { action: 'onOpen' },
    onClose: { action: 'onClose' },
    id: storybook.args.disabled
  }
} as Meta

const AlertButton: React.FC<AlertProps> = props => {
  const { message, options } = props
  const alert = useAlert()

  const handleAlertShow = () => alert.show(message, options)

  return (
    <Button onClick={handleAlertShow} size={'sm'}>
      Show
    </Button>
  )
}

const AlertStory: Story<AlertStoryProps> = args => {
  const { id, style, message, close, type, onOpen, onClose, ...alertProviderProps } = args

  const alertOptions = {
    id,
    style,
    message,
    close,
    options: {
      type,
      onOpen,
      onClose,
      position: alertProviderProps.position
    }
  }

  return (
    <AlertProvider template={Alert} {...alertProviderProps}>
      <AlertButton {...alertOptions} />
    </AlertProvider>
  )
}

AlertStory.args = {
  id: 'story',
  position: ALERT_PROVIDER_POSITIONS[4],
  timeout: 2000,
  offset: '0px 0px 0px 0px',
  transition: transitions.SCALE,
  style: {},
  type: 'success',
  message: 'Alert',
  onOpen: action('onOpen'),
  onClose: action('onClose')
}

export const SuccessAlert = AlertStory.bind({})
SuccessAlert.args = {
  ...AlertStory.args,
  message: 'Success'
}

export const InfoAlert = AlertStory.bind({})
InfoAlert.args = {
  ...AlertStory.args,
  message: 'Info',
  type: 'info'
}

export const ErrorAlert = AlertStory.bind({})
ErrorAlert.args = {
  ...AlertStory.args,
  message: 'Error',
  type: 'error'
}
