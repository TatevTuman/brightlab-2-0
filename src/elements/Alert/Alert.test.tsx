import React from 'react'
import { render, cleanup, waitFor } from '@testing-library/react'
import Alert from './Alert'

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('Alert', () => {
  const props = {
    id: 'test',
    message: 'Test message',
    options: {},
    close: jest.fn(),
    style: {}
  }

  it('renders correctly', async () => {
    const { container, getByText } = render(<Alert {...props} />)
    const awaitedAlert = await waitFor(() => container)

    // TODO
    expect(awaitedAlert).toMatchSnapshot()
    expect(getByText(props.message)).toBeInTheDocument()
  })
})
