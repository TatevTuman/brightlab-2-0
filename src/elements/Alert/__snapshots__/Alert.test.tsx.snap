// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert renders correctly 1`] = `
<div>
  <div
    class="alert"
  >
    Test message
  </div>
</div>
`;

exports[`Alert renders correctly 2`] = `
<div>
  <div
    class="alert"
  >
    Test message
  </div>
</div>
`;
